samza-api/src/main/java/org/apache/samza/operators/descriptors/base/stream/StreamDescriptor.java

```diff
@@ -0,0 +1,147 @@
+  private final Optional<SystemDescriptor> systemDescriptorOptional;
+
+  private Optional<String> physicalNameOptional = Optional.empty();
+  private Map<String, String> streamConfigs = new HashMap<>();
+
+  /**
+   * Constructs a {@link StreamDescriptor} instance.
+   *
+   * @param streamId id of the stream
+   * @param systemName system name for the stream
+   * @param serde serde for messages in the stream
+   * @param systemDescriptor system descriptor this stream descriptor was obtained from if available, else null
+   */
+  StreamDescriptor(String streamId, String systemName, Serde<StreamMessageType> serde,
+      SystemDescriptor systemDescriptor) {
+    this.streamId = streamId;
+    this.systemName = systemName;
+    if (serde == null) {
+      throw new SamzaException(
+          String.format("Serde must not be null for stream: %s on system: %s", streamId, systemName));
+    }
+    this.serde = serde;
+    if (systemDescriptor != null) {
+      if (!systemDescriptor.getSystemName().equals(systemName)) {
+        throw new SamzaException(
+            String.format("System name in constructor: %s does not match system name in SystemDescriptor: %s",
+                systemName, systemDescriptor.getSystemName()));
+      }
+      this.systemDescriptorOptional = Optional.of(systemDescriptor);
+    } else {
+      this.systemDescriptorOptional = Optional.empty();
+    }
+  }
+
+  /**
+   * The physical name of the stream on the system on which this stream will be accessed.
+   * This is opposed to the {@code streamId} which is the logical name that Samza uses to identify the stream.
+   * <p>
+   * A physical name could be a Kafka topic name, an HDFS file URN, or any other system-specific identifier.
+   * <p>
+   * If not provided, the logical {@code streamId} is used as the physical name.
+   *
+   * @param physicalName physical name for this stream.
+   * @return this stream descriptor.
+   */
+  public SubClass withPhysicalName(String physicalName) {
+    this.physicalNameOptional = Optional.ofNullable(physicalName);
+    return (SubClass) this;
+  }
+
+  /**
+   * Additional system-specific properties for this stream.
+   * <p>
+   * These properties are added under the {@code streams.stream-id.*} scope.
+   *
+   * @param streamConfigs system-specific properties for this stream
+   * @return this stream descriptor
+   */
+  public SubClass withStreamConfigs(Map<String, String> streamConfigs) {
+    this.streamConfigs.putAll(streamConfigs);
+    return (SubClass) this;
+  }
+
+  public String getStreamId() {
+    return this.streamId;
+  }
+
+  public String getSystemName() {
+    return this.systemName;
+  }
+
+  /**
+   * Get the serde for this stream. This can be the stream level serde if one was provided,
+   * or the default system level serde.
+   *
+   * @return the serde for this stream
+   */
+  public Serde getSerde() {
+    return this.serde;
+  }
+
+  public Optional<SystemDescriptor> getSystemDescriptor() {
```

Comment:
I was wondering if this class even needed a getSystemDescriptor at all. It is my understanding is that this is needed to find default fallback properties, so I was wondering if this could fully encapsulate those default fallbacks. Then, code would not need to worry about calling anything other than the stream descriptor.
This current solution is reasonable, but I just wanted to ask about this alternative.

Reply:
Sorry, not sure I understand what the suggestion is, can you clarify? We can discuss offline if that helps.

Here's an overview of how things currently work: A single SystemDescriptor can be associated with multiple input/output streams in the DAG. SystemDescriptors configure not just the fallbacks, but also properties of the system itself (e.g. Kafka bootstrap server url, ZK connect url). StreamDescriptors track what system descriptor they're associated with. In StreamGraph, we track the system descriptors separately from the input/output stream descriptors, and in ApplicationRunner we generate configs for them separately. This reflects how the configs are currently structured.

Note that this not meant to be a user facing API, it's for the framework's internal use. I'll try to separate the two more clearly, either using separate interfaces, or by documenting this more clearly. Would that address your concern?